erc20-migration/contracts/ZTESTZendBackupVault.sol

```diff
@@ -80,42 +97,141 @@ contract ZTESTZendBackupVault is Ownable {
         bytes32 messageHash = VerificationLibrary.createMessageHash(strMessageToSign);
         VerificationLibrary.verifyZendSignature(messageHash, signature, pubKeyX, pubKeyY);
 
-        balances[zenAddress] = 0;
-        zenToken.mint(destAddress, amount);
-        emit Claimed(destAddress, zenAddress, amount);   
+        _claim(destAddress, zenAddress);
+    }
+    
+    /// @notice Claim a P2SH balance.
+    ///         destAddress is the receiver of the funds
+    ///         hexSignatures is the array of the signatures of the claiming message. Must be generated in a compressed format if the public keys in the script are in compressed format, or uncompressed otherwise.
+    ///
+    ///         IMPORTANT: the array should have as length the number of public keys in the script. The signature in the "i" position should be the signature for the "i"
+    ///         pub key in the order it appears in the script. If the signature is not present for that key, it should be empty.
+    ///         This is to avoid duplicated signatures without expensive checks.
+    ///         
+    ///         script is the script to claim, from which pubKeys will be extracted
+    ///         pubKeysX and pubKeysY are the first 32 bytes and second 32 bytes of the signing keys for each one in the script (we use always the uncompressed format here)
+    ///         If the signature is not present for that key, the pub keys x and y should be bytes32(0)
+    ///         (Claiming message is predefined and composed by the string 'ZENCLAIM' concatenated with the zenAddress and destAddress in lowercase string hex format)
+    ///         (zenAddress is the string representation with 0x prefix )
+    function claimP2SH(address destAddress, bytes[] memory hexSignatures, bytes memory script, bytes32[] memory pubKeysX, bytes32[] memory pubKeysY) public canClaim(destAddress) {
+
+        uint256 minSignatures = uint256(uint8(script[0])) - 80;
+        _verifyPubKeysFromScript(script, pubKeysX, pubKeysY);
+        if(hexSignatures.length != pubKeysX.length) revert InvalidSignatureArrayLength(); //check method doc
+        bytes20 zenAddress = _extractZenAddressFromScript(script);
+
+        //address in signed message should respect EIP-55 format (https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md)
+        string memory destAddressAsString = Strings.toChecksumHexString(destAddress);
+        string memory zenAddressAsString = Strings.toHexString(address(zenAddress));
+        string memory strMessageToSign = string(abi.encodePacked(MESSAGE_PREFIX, zenAddressAsString, destAddressAsString));
+        bytes32 messageHash = VerificationLibrary.createMessageHash(strMessageToSign);
+
+        //check signatures
+        uint256 validSignatures;
+        uint256 i;
+        while(i != hexSignatures.length && validSignatures < minSignatures) {
+            if(pubKeysX[i] != bytes32(0) && hexSignatures[i].length != 0) {
```

Comment:
Considering that there are valid pub keys that have pubKeysX = 0, here it can be used pubKeysY instead